File: Graph/java/PrereqsPossible.java
import java.util.ArrayList;
import java.util.HashMap;
import java.util.HashSet;
import java.util.List;
import java.util.Map;
import java.util.Set;

public class PrereqsPossible {
  public static boolean possible(int numCourses, List<int[]> prereqs) {
    Map<Integer, List<Integer>> graph = new HashMap<>();
    for (int i = 0; i < numCourses; i++) {
      graph.put(i, new ArrayList<>());
    }
    for (int[] edge : prereqs) {
      graph.get(edge[0]).add(edge[1]);
    }

    Set<Integer> visiting = new HashSet<>();
    Set<Integer> visited = new HashSet<>();

    for (int i = 0; i < numCourses; i++) {
      if (cycle(graph, i, visiting, visited)) {
        return false;
      }
    }

    return true;
  }

  private static boolean cycle(Map<Integer, List<Integer>> graph, int node, Set<Integer> visiting, Set<Integer> visited) {
    if (visited.contains(node)) {
      return false;
    }
    if (visiting.contains(node)) {
      return true;
    }

    visiting.add(node);
    for (int neighbor : graph.get(node)) {
      if (cycle(graph, neighbor, visiting, visited)) {
        return true;
      }
    }

    visiting.remove(node);
    visited.add(node);
    return false;
  }
}

File: Graph/python/prereqsPossible.py
from typing import Dict, List, Set, Tuple


def prereqs_possible(num_courses: int, prereqs: List[Tuple[int, int]]) -> bool:
    graph: Dict[int, List[int]] = {i: [] for i in range(num_courses)}
    for a, b in prereqs:
        graph[a].append(b)

    visiting: Set[int] = set()
    visited: Set[int] = set()

    for node in range(num_courses):
        if _has_cycle(graph, node, visiting, visited):
            return False

    return True


def _has_cycle(graph: Dict[int, List[int]], node: int, visiting: Set[int], visited: Set[int]) -> bool:
    if node in visited:
        return False
    if node in visiting:
        return True

    visiting.add(node)
    for neighbor in graph[node]:
        if _has_cycle(graph, neighbor, visiting, visited):
            return True
    visiting.remove(node)
    visited.add(node)
    return False
